python/ray/experimental/channel/shared_memory_channel.py

```diff
@@ -380,3 +381,89 @@ def close(self) -> None:
         if self.is_local_node(self._reader_node_id):
             self.ensure_registered_as_reader()
         self._worker.core_worker.experimental_channel_set_error(self._reader_ref)
+
+
+@PublicAPI(stability="alpha")
+class MultiChannel(ChannelInterface):
+    def __init__(
+        self,
+        writer: Optional[ray.actor.ActorHandle],
+        readers: List[Optional[ray.actor.ActorHandle]],
+        _local_channel: Optional[LocalChannel] = None,
+        _remote_channel: Optional[Channel] = None,
+    ):
+        """
+        Can be used to send data to different readers via different channels.
+        For example, if the reader is in the same worker process as the writer,
+        the data can be sent via LocalChannel. If the reader is in a different
+        worker process, the data can be sent via shared memory channel.
+        """
+        self._writer = writer
+        self._readers = readers
+        self._local_channel = _local_channel
+        self._remote_channel = _remote_channel
+
+        remote_readers = []
+        for reader in self._readers:
+            if self._writer != reader:
+                remote_readers.append(reader)
+        # There are some local readers which are the same Ray actor as the writer.
+        # Create a local channel for the writer and the local readers.
+        if not self._local_channel and len(remote_readers) != len(self._readers):
+            self._local_channel = LocalChannel(self._writer)
+        # There are some remote readers which are not the same Ray actor as the writer.
+        # Create a shared memory channel for the writer and the remote readers.
+        if not self._remote_channel and len(remote_readers) != 0:
+            self._remote_channel = Channel(self._writer, remote_readers)
+        assert hasattr(self, "_local_channel") or hasattr(self, "_remote_channel")
+
+    def ensure_registered_as_writer(self) -> None:
+        if self._local_channel:
+            self._local_channel.ensure_registered_as_writer()
+        if self._remote_channel:
+            self._remote_channel.ensure_registered_as_writer()
+
+    def ensure_registered_as_reader(self) -> None:
+        if self._local_channel:
+            self._local_channel.ensure_registered_as_reader()
+        if self._remote_channel:
+            self._remote_channel.ensure_registered_as_reader()
+
+    def __reduce__(self):
+        return MultiChannel, (
+            self._writer,
+            self._readers,
+            self._local_channel,
+            self._remote_channel,
+        )
+
+    def write(self, value: Any):
+        if self._local_channel:
+            self._local_channel.write(value)
+        if self._remote_channel:
+            self._remote_channel.write(value)
+
+    def use_local_channel(self):
```

Comment:
Suggest replacing this with a flag that's set in the constructor, since we don't expect the return value to ever change once the channel has been deserialized.

Reply:
`use_local_channel` has already been removed. We maintain a channel dict instead.